Rework warning 53 (misplaced attributes) to work on all attributes in all places

ccasin opened the pull request:
This PR reworks warning 53 (misplaced attribute) so that it works for all attributes and is issued in a uniform way.

Presently, there are many attributes known to the compiler for which this warning is never issued (even though there are many places one can put them that they will be ignored).  Those for which the warning can be issued are handled by just looking for them in specific places we think someone might mistakenly write them.  The goal of this PR is to fix that situation.

The main idea is to take an approach like ppxlib: When the parser sees an attribute, it adds it to a table.  When the compiler uses the attribute, it marks it as used by removing it from the table.  When compilation is done, we check whether any attributes remain in the table, and issue warnings for them.

We only track known compiler built-in attributes for the warning.  There is now a list of such attributes in `Builtin_attributes`.  While most ppxes now remove their attributes after processing them, not all do, and other attributes may be in the program for various reasons.  To deal with the problem that the compiler can't know whether non-built-in attributes are misplaced, we only track those attributes which appear in the list of built-ins.

When we deployed this within Jane Street, we found more than 200 misplaced attributes in our code base.  And this PR even cleans up a few misplaced attributes in the stdlib.  So, I believe this is a useful addition.

## Reviewing

This PR isn't massive, but I've broken it into several commits to create an opportunity to explain a few subtleties:

### Commit 1: Add attr tracking mechanism, but don't whitelist any attrs yet

This adds all the new attribute tracking code, but leaves the list of tracked attributes empty and the old w53 stuff in place.  So all the code to add attributes to the table is there and fires, but no attributes are on the whitelist so they aren't added.

A couple notes about when attributes are added to the table:

**When to add attributes**: It's not sufficient to just add attributes when they are parsed in the parser, because not all parsetrees come from the parser.  The `-pp` flag will read a marshalled parsetree instead, and ppxes may be used to modify the parsetree.

We want to track only attributes in the final version of the parsetree.  We can accomplish this by adding attributes to the table in two cases:

- If no `-ppx` or `-pp` flags are passed, the parser will be run and the parsetree won't be modified, so we add attributes during parsing.
- If either of those flags is passed, the `Ast_invariants` check will be run on the final parsetree, so we add attributes during that check instead.

**Attributes in attributes**:  Attribute payloads can contain more attributes.  [This happens in practice](https://github.com/LexiFi/gen_js_api/blob/d91966229284995f179d1889c822e57853f6267e/examples/misc/jquery.mli#L8), if rarely.

These inner attributes should not be tracked for w53 - we can't know the meaning of the outer attribute, so we have no way to know what to do with them and whether or not they were used appropriately.

If we're adding attributes during the `Ast_invariants` check, that's easy - just update the iterator when descending into an attribute payload.  In the parser, it's harder.  I've done the dumb thing of just removing any nested attributes from the table after parsing an attribute.  This wastes a little effort, but attribute payloads are typically small and nested attributes are rare, so it shouldn't be too bad.  This strategy seemed much simpler and easier to review than trying to thread some state through the parser, but I'm open to other suggestions.

### Commit 2: Rework w53 tests

This replaces the w53 test with a new one that tests all compiler built-in attributes.  For each attribute, I tried to include examples of where it is allowed to appear and where it is not and w53 is issued (if such a place exists). The corresponding reference file is updated over the next two commits as attributes are added to the whitelist.

### Commit 3: Handle the basic attributes

This adds most attributes for the whitelist, leaving off those related to warnings (which need to be handled a little differently).

The bulk of the diff here is in `lambda/translattribute.ml`.  There are several related changes:
- We should use the accessors in `Builtin_attributes` to look for specific attributes, ensuring they are marked as used.
- A few functions, like `add_local_attribute`, are slightly rearranged so that attributes aren't checked for until we're sure we're in a context where they can be used (since the act of looking for the attribute marks it as used).
- It's no longer necessary to remove these attributes from the terms when they are used.  The old implementation of w53 for those attributes, which relied on checking whether attributes remained on terms after translation, is removed.
- Care must be taken to mark an attribute as used even if we're compiling with a configuration that does not use it.  E.g., the `unrolled` attribute only has an effect on the program if compiling with `flambda`, but should not issue `w53` when compiling with `closure`.

This commit fixes a bug in `find_attribute`, I think: It used to be that if an attribute was written twice on the same term (e.g., `[@inline] [@inline]`), it was removed with a warning.  Now one of them is used, with a warning.

It also removes a few misplaced `[@inline]` attributes in the stdlib (these attributes do nothing in mli files).

### Commit 5: Handle attributes like "warning" and "alert"

Attributes like `[@alert]` are "used" by placing them in the environment, where they may be noticed later.  So this commit adds code to `env.ml` to mark such attributes used when they are added to the environment as part of some form that is allowed to use them.  The special case of a file-level `[@@@alert]` is handled in `typemod.ml`.

In the stdlib there is an `[@@@alert]` attribute in the hashtable template file which is particularly annoying:  That template is included in one place at the top level of a file (so the attribute is legal) and another place inside a module (where the attribute has no effect).  I've dealt with this by using warning attributes to disable w53 here.

### Commit 6: `[@ppwarning]` with a bad payload should give w47, not w53

This just fixes a minor issue I noticed in the final state of this PR.  `[@ppwarning]` with a bad payload used to be silently ignored, but after this PR it would issue w53.  But warning 47 (bad attribute payload) is more appropriate.

ccasin committed:
Add attr tracking mechanism, but don't whitelist any attrs yet

ccasin committed:
Rework w53 tests.

The reference file is not correct yet - it will be updated in subsequent commits
as we track attributes with the new mechanism.

ccasin committed:
[@ppwarning] with a bad payload should give w47, not w53

ccasin committed:
changes entry

ccasin commented:
~~There is a problem here with the stdlib template mechanism (`hashtbl.template.mli` has an attribute that works fine in `hashtbl.mli` but is incorrect when that signature is inlined in `moreLabels.mli`, because `[@@@alert]` only works at the file level).~~

~~I don't know much about the stdlib template mechanism, so I'm not immediately sure what the right fix is.  Happy for suggestions, or I'll look closer tomorrow.~~

(Now fixed, see below)

ccasin committed:
Handle the basic attributes.

Needs a bootstrap due to removing misplaced attrs in the stdlib.  That's done in
next commit.

ccasin committed:
bootstrap, promote parser, make depend

ccasin committed:
Handle attributes like "warning" and "alert"

This also makes w53 controllable with warning attributes, but only for top-level
alerts.  This is used to handle the fact that in the stdlib there's a template
file with a top-level alert attribute, and it's included in two different
places, one where that attribute is legal, and one where it is not

A boostrap is needed, and performed in the next commit.

ccasin committed:
bootstrap

ccasin commented:
OK, I've (hopefully) fixed the stdlib / templates issue by making warning 53 for top-level alerts controllable with warning attributes, and turning it off around the relevant attribute in the template file.  It would be nice to make it fully controllable with warning attributes, but that seems a bit tricky and best left for a subsequent PR.  Though I'm open to other suggestions regarding the template file issue.

I've rebased, adjusted the history, and force pushed, putting this fix as part of the commit that dealt with alert attributes and cleaning the bootstrapping into its own commits.  If anyone wants to see just the fix added for the template issue, I put it [here](https://github.com/ocaml/ocaml/commit/046a44c3de862b25c2b91743722e455dd668362a).  I've adjusted the original PR description to reflect this.

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -15,6 +15,72 @@
+
+  let hash : t -> int = Hashtbl.hash
+  let equal : t -> t -> bool = (=)
+end)
```
This looks like a mutable set with an order (the `attr_order` below) on the elements, was there a specific reason to chose an `Hashtbl` ?

ccasin replied:
This seemed like the most efficient choice, since the stdlib lacks a mutable set.  `ppxlib` makes the same choice for [their similar structure](https://github.com/ocaml-ppx/ppxlib/blob/af1756cfcfb6ebabd2ebf2234109cd50fce3f2f8/src/attribute.ml#L307).  But what would you recommend?

Octachron replied:
I was thinking of using a `Attribute_set ref` with the attribute set instantiated with a total version of `attr_order` to avoid the `sort` function. But a hashtbl as mutable set is fine.

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -15,6 +15,72 @@
+  ; "warnerror"; "ocaml.warnerror"
+  ; "warning"; "ocaml.warning"
+  ; "warn_on_literal_pattern"; "ocaml.warn_on_literal_pattern"
+  ]
```
Rather than listing each built-in in its prefixed and non-prefixed form, I propose to move the prefixing logic to the lookup functions.

ccasin replied:
I've made this change, and the similar suggested changes below.  This resulted in some additional simplifications (various functions that used to take lists can now just take a single attribute name).

I was on the fence, after reading the suggestion, because a consequence of this move is that we'll needlessly allocate additional strings when we strip `ocaml.` prefixes.  However, after seeing the simplifications, I'm happy with the change.  (Also, for small programs or programs with no attributes, this is probably just a win.)

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -67,6 +133,40 @@ let error_of_extension ext =
 
+let mark_alert_used a =
+  match a.attr_name.txt with
+  | "ocaml.deprecated"|"deprecated"|"ocaml.alert"|"alert" ->
```
Similarly, do we want to add an `unprefix` function that removes the `ocaml.` prefix to avoid the duplication here?

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -243,34 +351,36 @@ let warning_scope ?ppwarning attrs f =
 
-
-let warn_on_literal_pattern =
+let has_attribute nms attrs =
```
As far as I can see, all arguments of `has_attribute` have the form `["ocaml." ^ attribute, attribute]` which suggest to move the namespacing logic inside the `has_attribute` function.

Octachron commented on parsing/builtin_attributes.mli:
```diff
@@ -13,26 +13,83 @@
+    - ocaml.inline
+    - ocaml.inlined
+    - ocaml.noalloc
+    - ocaml.ppwarning
```
`ocaml.poll` is missing in this list (since before this PR).

ccasin replied:
Indeed: Also missing were `specialise`, `specialised` and `tail_mod_cons`.  I have added these and corresponding tests.

I'll also note that all four are missing from the [manual page](https://v2.ocaml.org/manual/attributes.html#ss%3Abuiltin-attributes) on built-in attributes.  I was tempted to fix that in this PR, but it's already a decent size and I'm probably not the best person to document their behavior.

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -15,6 +15,72 @@
+  | Parser when !Clflags.all_ppx <> [] -> ()
+  | Parser | Invariant_check ->
+    if is_builtin_attr name.txt then
+      Attribute_table.replace unused_attrs name ()
```
The use of `replace` here means that we lose the duplicated attribute warning for attributes added by a ppx to the same location, isn't it?

ccasin replied:
I think there are a couple ways to read this remark, depending on whether you meant (1) "(duplicate attribute) warning" or (2) "duplicate (attribute warning)" .  Let me answer both :)

1. The logic for `Warnings.Duplicated_attribute` (warning 54) is meant to be unchanged by this PR.  In particular, it doesn't use this table at all - it looks for multiple copies of the attribute on the term, which will still be there.  So I think there is no change.

2. The other question would be how many times we issue warning 53 if a ppx adds multiple copies of an attribute with the same location in an incorrect spot.  I believe the old answer is 0 (even in cases where w53 otherwise worked well) and the new answer is 1.  I say the old answer is 0 because the old implementation looked for attributes using [find_attribute](https://github.com/ocaml/ocaml/blob/f6d6f65f0fe985101e7455a60a025a693a5f212e/lambda/translattribute.ml#L49), which would return no attributes at all in the case of a duplicate (this is the small bug I mentioned fixing in commit 3).

As for point 2, we could imagine issuing warning 53 multiple times for the same location by recording in the hashtbl how many times this attribute has been seen at this location.  Or adjusting the one warning to mention multiple copies were found, with the same technique.  This does not seem worth doing, to me, for the corner case of a broken ppx - but I'm happy to add it if it seems useful to you.

Octachron replied:
I was thinking of case 2, but I agree that having one warning emitted is probably the right behavior.

Octachron commented on parsing/builtin_attributes.mli:
```diff
@@ -13,26 +13,83 @@
+    attributes in that case.
+*)
+type attr_tracking_time = Parser | Invariant_check
+val register_attr : attr_tracking_time -> string Location.loc -> unit
```
I find the documentation of  `attr_tracking_time` argument confusing, since the argument is here to indicate in which phase of the pipeline we are calling the function (during parsing or during the ast invariant check), and which is unrelated to the attribute `attr` at hand.

ccasin replied:
I've changed the name (as you suggested below) and adjusted its documentation.

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -15,6 +15,72 @@
+
+let is_builtin_attr s = Hashtbl.mem builtin_attrs s
+
+type attr_tracking_time = Parser | Invariant_check
```
Isn't that more a `current_phase` rather than a `tracking time` attached to the specific attribute?

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -243,34 +351,36 @@ let warning_scope ?ppwarning attrs f =
+    attrs
+
+type attr_action = Mark_used_only | Return
+let filter_attributes actions attrs =
```
`filter` or `select`? I am also wondering if we should focus on the action on the attribute or the status (`Active | Inactive` ) of the attribute  (with the same semantic).

ccasin replied:
I agree select is clearer and have changed it.

Octachron commented on parsing/builtin_attributes.mli:
```diff
@@ -13,26 +13,83 @@
+(** Marks the attribute used for the purposes of misplaced attribute warnings if
+    it is an alert.  Call this when moving things allowed to have alert
+    attributes into the environment. *)
+val mark_alert_used : Parsetree.attribute -> unit
```
I would propose to place those four `mark_` function in a separate documentation section and add a common paragraph that those functions are necessary because we sometimes move definitions with raw attributes in the environment, and only check the attributes on a potential warning/alert.

ccasin replied:
Done.  I added sections to the whole file - hopefully they make sense.

Octachron reviewed (commented):
I have some minor remarks on the `Builtin_attributes` interface with respect to the `ocaml` namespace for attributes, but otherwise that PRs feels like a clear improvement to me.

Marking built-in attributes as we use them should be far less error prone than trying to remember to check if some builtin attributes were left unused somewhere or somewhen.

ccasin commented:
@Octachron Thanks very much for the thoughtful review, and apologies for the delay in getting back to this.  I've responded to a couple points above.  I meant to rebase and integrate the review feedback today as well, but got distracted by tracking down #12697 and am now out of time.  I will return later in the week to integrate your suggestions.

ccasin committed:
integrate review: move namespace logic to eliminate duplication

This also simplifies the types of several functions, which no longer need to
take lists.

ccasin committed:
integrate review: rename [attr_tracking_time] and [filter_attributes]

ccasin committed:
integrate review: doc sections in builtin_attributes

ccasin committed:
integrate review: missing attributes

ccasin commented:
@Octachron I believe I have addressed all your comments above with four new commits, and this is ready for further review.  I've also rebased it on to the latest trunk.  Thanks again!

Octachron commented on parsing/builtin_attributes.ml:
```diff
@@ -15,6 +15,83 @@
+  if String.starts_with ~prefix:"ocaml." s && len > 6 then
+    String.sub s 6 (len - 6)
+  else
+    s
```
Note that it seems that in a handful of cases, we could avoid the allocation by having a `attr_equal` function rather than using the `String.equal (drop_ocaml_attr_prefix x) y` pattern.

ccasin replied:
Good idea - I've made this change, leaving only the allocation in `is_builtin_attr`.  This seems like a fine compromise.

Octachron commented on testsuite/tests/warnings/w53.ml:
```diff
@@ -1,57 +1,507 @@
+end
+
+module TestDeprecatedStruct = struct  (* CJC XXX THIS IS ALL BUGGY *)
+  let x = 5 [@deprecated] (* rejected *)
```
What is buggy in this test?

ccasin replied:
Oops, sorry, nice catch!  This was a stray comment from an intermediate state, now deleted.

Octachron reviewed (approved):
The PR looks good to merge to me (except for some minor remarks).
I really like the fact that with this PR the misplaced attribute warning will only require care from attribute authors.

ccasin committed:
integrate review: allocate fewer strings and delete a stray comment

ccasin commented:
I've integrated the last round of minor feedback, and believe this is good to go when the tests pass.

gasche commented:
What should we do with the git history? Squash everything, something else?

ccasin commented:
> What should we do with the git history? Squash everything, something else?

I think squashing is reasonable - it should read fine as one commit.  But I'm open to suggestions if you'd like me to reorganize it.

Octachron commented:
The commit history seems mostly linear (no back-and-forth) and contains some bootstrap, thus I think it is clearer to merge the PR as it is.
